File: scripts/watermark_remove.py
#!/../utility-tools-env/Scripts/python
import sys
import os
from PIL import Image
import time


def remove_watermark(img_path, result_path, grayscale_min, grayscale_max):
    if not os.path.exists(result_path):
        os.makedirs(result_path)

    image_files = [file for file in os.listdir(
        img_path) if file.endswith('.jpg')]

    for file in image_files:
        image_path = os.path.join(img_path, file)
        image = Image.open(image_path).convert('L')
        image = image.point(lambda p: 255 if p >
                            grayscale_max else 0 if p < grayscale_min else p)

        new_image_path = os.path.join(result_path, file)
        image.save(new_image_path)


if len(sys.argv) < 5:
    print("Usage: python watermark_remove.py <input_path> <output_path> <grayscale_min> <grayscale_max>")
    sys.exit(1)

img_path = sys.argv[1]
result_path = sys.argv[2]
grayscale_min = int(sys.argv[3])
grayscale_max = int(sys.argv[4])

remove_watermark(img_path, result_path, grayscale_min, grayscale_max)

File: routes/watermark-remove.js
const { exec } = require('child_process');
const multer = require('multer');
const fs = require('fs');
const path = require('path');
const scriptPath = path.join(__dirname, '..', 'scripts', 'watermark_remove.py');
const router = require('express').Router();

const storage = multer.diskStorage({
    destination: function (req, file, cb) {
        cb(null, 'public/image-cache');
    },
    filename: function (req, file, cb) {
        const decodedFileName = decodeURIComponent(file.originalname);
        cb(null, decodedFileName);
    },
});
const upload = multer({ storage: storage });

function removeWatermark(inputPath, outputPath, grayscaleMin, grayscaleMax) {
    return new Promise((resolve, reject) => {
        const command = `python ${scriptPath} ${inputPath} ${outputPath} ${grayscaleMin} ${grayscaleMax}`;
        exec(command, (error, stdout, stderr) => {
            if (error) {
                reject(error);
            } else {
                const processedImages = getProcessedImages(outputPath);
                resolve({ images: processedImages });
            }
        });
    });
}

function getProcessedImages(outputPath) {
    const imageFiles = fs.readdirSync(outputPath).filter((file) => file.endsWith('.jpg'));
    const processedImages = imageFiles.map((file) => `/image-result-cache/${file}`);
    return processedImages;
}

router.post('/', upload.array('images'), (req, res) => {
    const grayscaleMin = req.body.minGrayscale;
    const grayscaleMax = req.body.maxGrayscale;
    const inputPath = 'public/image-cache';
    const outputPath = 'public/image-result-cache';
    removeWatermark(inputPath, outputPath, grayscaleMin, grayscaleMax)
        .then(result => {
            res.json(result);
        })
        .catch(error => {
            console.error('Watermark removal error:', error);
            res.status(500).json({ error: 'Watermark removal failed' });
        });
});

module.exports = router;
